### PR TITLE
Load the nested models of eager loaded models

### PR DESCRIPTION
**Pull request by jcmfernandes**
Generated nested models register themselves upon being created by
`Avromatic::Model::Builder`. If Rails's autoloader didn't recreate a model that
should be eager loaded, then its nested models won't be registered as we hook on
Railtie `#to_prepare` that itself triggers a clean-up of the nested models
registry.

Therefore, we have to go through the attribute definitions of each eager loaded
model and try to register its nested models.

prime @jturkel

### CODE REVIEW

File: lib/avromatic.rb [hunk 1]
```diff
@@ -77,19 +77,38 @@ def self.prepare!(skip_clear: false)
...
     end
   end
   private_class_method :eager_load_models!
+
+  def self.eager_load_nested_models!(type)
+    case type
+    when Avromatic::Model::Types::ArrayType, Avromatic::Model::Types::MapType
+      eager_load_nested_models!(type.value_type)
+    when Avromatic::Model::Types::CustomType
+      eager_load_nested_models!(type.default_type) # correct?
```
Comment (jcmfernandes) [inline lib/avromatic.rb]: @jturkel I assume this makes sense but let me know.

File: lib/avromatic.rb [hunk 2]
```diff
@@ -77,19 +77,38 @@ def self.prepare!(skip_clear: false)
...
+    when Avromatic::Model::Types::UnionType
+      type.member_types.each { |member_type| eager_load_nested_models!(member_type) }
+    when Avromatic::Model::Types::RecordType
+      nested_models.ensure_registered_model(type.record_class)
```
Comment (jcmfernandes) [inline lib/avromatic.rb]: Begs the question if we should restart the recursion with the `#attribute_definitions` of `type.record_class`.

File: lib/avromatic.rb [hunk 3]
```diff
@@ -77,19 +77,38 @@ def self.prepare!(skip_clear: false)
...
   def self.eager_load_models=(models)
     @eager_load_model_names = Array(models).map { |model| model.is_a?(Class) ? model.name : model }
   end
 
-  def self.eager_load_models!
+  def self.eager_load_models!(skip_nested_models: true)
     (@eager_load_model_names || []).each do |model_name|
-      nested_models.ensure_registered_model(model_name.constantize)
+      model = model_name.constantize
+      next if !nested_models.ensure_registered_model(model) || skip_nested_models
```
Comment (jturkel) [inline lib/avromatic.rb]: Relying on `ensure_registered_model` to return a truthy value is a bit dangerous given the current implementation (i.e. it works today but likely was not intentional). Adding a `next if nested_models.registered?(model)` guard is safer.

File: lib/avromatic.rb (hunk 1 above)
Comment (jturkel) [reply lib/avromatic.rb]: Looks correct to me!

File: lib/avromatic.rb [hunk 4]
```diff
@@ -77,19 +77,38 @@ def self.prepare!(skip_clear: false)
...
+
+      model.attribute_definitions.each do |_attribute, definition|
+        eager_load_nested_models!(definition.type)
+      end
     end
   end
   private_class_method :eager_load_models!
+
+  def self.eager_load_nested_models!(type)
+    case type
```
[the remark this answers is not in the document]
Comment (jcmfernandes) [reply lib/avromatic.rb]: Addressed.

File: lib/avromatic.rb [hunk 5]
```diff
@@ -77,19 +77,38 @@ def self.prepare!(skip_clear: false)
         end
       end
     end
-    eager_load_models!
+    eager_load_models!(skip_nested_models: skip_clear)
```
[the remark this answers is not in the document]
Comment (jcmfernandes) [reply lib/avromatic.rb]: Addressed,

**Comment by jcmfernandes** [comment]
Hey @jturkel, I did some changes based on your feedback. Let me know what you think. Thanks!

File: lib/avromatic.rb [hunk 6]
```diff
@@ -81,13 +82,11 @@ def self.prepare!(skip_clear: false)
   end
 
   def self.eager_load_models=(models)
-    @eager_load_model_names = Array(models).map { |model| model.is_a?(Class) ? model.name : model }
+    @eager_load_model_names = Array.wrap(models).map { |model| model.is_a?(Class) ? model.name : model }
```
Comment (jturkel) [inline lib/avromatic.rb]: Was this change necessary? `Array()` and `Array.wrap()` should behave the same here.

File: lib/avromatic/model/types/abstract_type.rb [hunk 7]
```diff
@@ -31,6 +31,10 @@ def coerced?(_value)
         def serialize(_value, **)
           raise "#{__method__} must be overridden by #{self.class.name}"
         end
+
+        def referenced_models
```
Comment (jturkel) [inline lib/avromatic/model/types/abstract_type.rb]: Should we call this `referenced_model_classes` to make it more obvious that we're not returning instantiated models?

File: lib/avromatic/model/types/map_type.rb [hunk 8]
```diff
@@ -68,6 +68,10 @@ def serialize(value, strict:)
             end
           end
         end
+
+        def referenced_models
+          [key_type, value_type].tap(&:uniq!).flat_map(&:referenced_models).tap(&:uniq!)
```
Comment (jturkel) [inline lib/avromatic/model/types/map_type.rb]: nit: Doesn't seem like the extra `tap(&:uniq!)` buys us much
```suggestion
          [key_type, value_type].flat_map(&:referenced_models).tap(&:uniq!)
```

File: lib/avromatic/model/types/boolean_type.rb [hunk 9]
```diff
@@ -33,6 +33,10 @@ def coercible?(input)
         def serialize(value, **)
           value
         end
+
+        def referenced_models
+          []
```
Comment (jturkel) [inline lib/avromatic/model/types/boolean_type.rb]: It might be worth declaring an `EMPTY_ARRAY` constant in `AbstractType` that subclasses can use. I don't expect a noticeable perf gain though since this method won't be called frequently. Up to you.

File: lib/avromatic/model/types/map_type.rb (hunk 8 above)
Comment (jcmfernandes) [reply lib/avromatic/model/types/map_type.rb]: It's my lazy way of performing the search twice for the same type. I'll leave it as-is if you don't have strong feelings about it.

File: lib/avromatic.rb (hunk 6 above)
Comment (jcmfernandes) [reply lib/avromatic.rb]: Not really, it was my OCD kicking in, as I have some fun stories to tell about `Array()` and it calling `#to_ary` first. But I can revert it 😅 .

**Comment by jcmfernandes** [comment]
Alright, all done! 😉

File: lib/avromatic/model/types/map_type.rb (hunk 8 above)
Comment (jturkel) [reply lib/avromatic/model/types/map_type.rb]: I don't have strong feelings about it. Avro [maps](https://avro.apache.org/docs/current/spec.html#Maps) always have string keys so `key_type.referenced_models` should always return the empty array anyway.

Comment (jcmfernandes) [reply lib/avromatic/model/types/map_type.rb]: Oh interesting! As a non-native speaker, it's my understanding that the spec doesn't force keys to be strings, but google says otherwise. https://github.com/apache/avro/pull/17/files even shows "non-string" maps being encoded as arrays to work around this.

I'm going to include this info in the PR.
